### PR TITLE
Added User entity, DTOs, AuthenticationController.

### PR DESCRIPTION
**Pull request by sovisou**

### CODE REVIEW

File: src/main/java/com/onlinebookstore/controller/AuthenticationController.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+public class AuthenticationController {
+    private final UserService userService;
+
+    @PostMapping("/registration")
+    public UserResponseDto registerUser(@RequestBody UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        return userService.register(requestDto);
+    }
+}
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/controller/AuthenticationController.java]:
```suggestion
```

File: src/main/java/com/onlinebookstore/controller/AuthenticationController.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/auth")
+@RequiredArgsConstructor
+public class AuthenticationController {
+    private final UserService userService;
+
+    @PostMapping("/registration")
+    public UserResponseDto registerUser(@RequestBody UserRegistrationRequestDto requestDto)
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/controller/AuthenticationController.java]: Valid annotation missed. Let’s  add open API documentation

File: src/main/java/com/onlinebookstore/exception/CustomGlobalExceptionHandler.java [hunk 3]
```diff
@@ -51,4 +51,13 @@ protected ResponseEntity<Object> entityNotFoundException(EntityNotFoundException
...
         return new ResponseEntity<>(body, HttpStatus.NOT_FOUND);
     }
+
+    @ExceptionHandler(RegistrationException.class)
+    protected ResponseEntity<Object> registrationException(RegistrationException ex) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.BAD_REQUEST);
+        body.put("message", ex.getMessage());
+        return new ResponseEntity<>(body, HttpStatus.BAD_REQUEST);
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/exception/CustomGlobalExceptionHandler.java]: I see similar logic in the previous method. Let’s move it to the private method ;)

File: src/main/java/com/onlinebookstore/exception/RegistrationException.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package com.onlinebookstore.exception;
+
+public class RegistrationException extends Exception {
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/exception/RegistrationException.java]:
```suggestion
```

File: src/main/java/com/onlinebookstore/mapper/UserMapper.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+import com.onlinebookstore.dto.user.UserResponseDto;
+import com.onlinebookstore.model.User;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
+public interface UserMapper {
+    UserResponseDto toUserResponse(User user);
+
+    User toModel(UserRegistrationRequestDto requestDto);
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/mapper/UserMapper.java]:
```suggestion
```

File: src/main/java/com/onlinebookstore/dto/user/UserResponseDto.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package com.onlinebookstore.dto.user;
+
+import lombok.Data;
+
+@Data
+public class UserResponseDto {
+    private Long id;
+    private String email;
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/dto/user/UserResponseDto.java:8]: Let’s  provide other fields (exclude password)

File: src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java]: better use existByEmail

File: src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            throw new RegistrationException("Can`t register user");
+        }
+        User user = new User();
+        user.setEmail(requestDto.getEmail());
+        user.setPassword(requestDto.getPassword());
+        user.setFirstName(requestDto.getFirstName());
+        user.setLastName(requestDto.getLastName());
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java]: Can you use a mapper for this logic?

File: src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        User savedUser = userRepository.save(user);
+        return userMapper.toUserResponse(savedUser);
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java]:
```suggestion
```

File: src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
+            throw new RegistrationException("Can`t register user");
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java]: Let’s add an email parameter to the exception message.

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+                  name: last_name
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: shipping_address
+                  type: varchar(255)
+
+
+
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml]: only one empty line at the end of the file.

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 12]
```diff
@@ -1,3 +1,6 @@
 databaseChangeLog:
   - include:
       file: db/changelog/changes/01-create-books-table.yaml
+  - include:
+      file: db/changelog/changes/02-create-users-table.yaml
+
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: the same

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution;)

File: src/main/java/com/onlinebookstore/dto/user/UserResponseDto.java (hunk 6 above)
Comment (Elena-Bruyako) [reply src/main/java/com/onlinebookstore/dto/user/UserResponseDto.java:8]: Comment not fixed
Example of response body:

```
{
  "id": 1,
  "email": "john.doe@example.com",
  "firstName": "John",
  "lastName": "Doe",
  "shippingAddress": "123 Main St, City, Country"
}
```

File: src/main/java/com/onlinebookstore/exception/CustomGlobalExceptionHandler.java [hunk 13]
```diff
@@ -45,10 +45,19 @@ private String getErrorMessage(ObjectError e) {
 
     @ExceptionHandler(EntityNotFoundException.class)
     protected ResponseEntity<Object> entityNotFoundException(EntityNotFoundException ex) {
+        return objectResponseEntity(HttpStatus.NOT_FOUND, ex.getMessage());
+    }
+
+    @ExceptionHandler(RegistrationException.class)
+    protected ResponseEntity<Object> registrationException(RegistrationException ex) {
+        return objectResponseEntity(HttpStatus.BAD_REQUEST, ex.getMessage());
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/exception/CustomGlobalExceptionHandler.java]:
```suggestion
        return objectResponseEntity(HttpStatus.CONFLICT, ex.getMessage());
```

File: src/main/java/com/onlinebookstore/model/User.java [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+@Setter
+public class User {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false, unique = true)
+    private String email;
+    @Column(nullable = false)
+    private String password;
+    @Column(name = "first_name", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/model/User.java]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java (hunk 7 above)
Comment (Elena-Bruyako) [reply src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java]: Comment not fixed
```suggestion
        if (userRepository.existsByEmail(requestDto.getEmail())) {
```

File: src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
+            throw new RegistrationException("Can`t register user with email: "
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java]: improve message like 
User with this email: %s already exists

File: src/main/java/com/onlinebookstore/validation/FieldMatchValidator.java [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+    private String secondFieldName;
+
+    @Override
+    public void initialize(FieldMatch constraintAnnotation) {
+        this.firstFieldName = constraintAnnotation.first();
+        this.secondFieldName = constraintAnnotation.second();
+    }
+
+    @Override
+    public boolean isValid(Object o, ConstraintValidatorContext constraintValidatorContext) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/validation/FieldMatchValidator.java]: what is it "o" ?

File: src/main/java/com/onlinebookstore/validation/FieldMatchValidator.java [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+        try {
+            Object firstValue = new BeanWrapperImpl(o).getPropertyValue(firstFieldName);
+            Object secondValue = new BeanWrapperImpl(o).getPropertyValue(secondFieldName);
+
+            return firstValue != null && firstValue.equals(secondValue);
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/validation/FieldMatchValidator.java]: you can use `Objects.equals`

File: src/main/java/com/onlinebookstore/validation/FieldMatchValidator.java [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+        } catch (Exception e) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/validation/FieldMatchValidator.java]: remove try - catch block

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+                  constraints:
+                    nullable: false
+              - column:
+                  name: last_name
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: shipping_address
+                  type: varchar(255)
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml:39]: add empty line at the end of file

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 20]
```diff
@@ -1,3 +1,5 @@
 databaseChangeLog:
   - include:
       file: db/changelog/changes/01-create-books-table.yaml
+  - include:
+      file: db/changelog/changes/02-create-users-table.yaml
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/db.changelog-master.yaml:5]: add empty line at the end of file

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml (hunk 19 above)
Comment (sovisou) [reply src/main/resources/db/changelog/changes/02-create-users-table.yaml:39]: There is a blank line, should I add one more?

Comment (liliia-ponomarenko) [reply src/main/resources/db/changelog/changes/02-create-users-table.yaml:39]: no, everything is fine now)

**Review by liliia-ponomarenko — APPROVED** [review]
Great! 😊

**Pull request merged**